### PR TITLE
MB-59978:

### PR DESCRIPTION
**Pull request by lileichen**
- Only allow mobile active if enableCrossClusterVersioning is true
- Raise warning if importCas is detected when mobile is off

### CODE REVIEW

File: parts/xmem_nozzle.go [hunk 1]
```diff
@@ -1340,6 +1343,14 @@ func (xmem *XmemNozzle) batchSetMetaWithRetry(batch *dataBatch, numOfRetry int)
 					ImportMutation: item.ImportMutation,
 				}
 				xmem.RaiseEvent(common.NewEvent(common.DataFailedCRSource, nil, xmem, nil, additionalInfo))
+				if item.ImportMutation && xmem.config.mobileCompatible == base.MobileCompatibilityOff && !xmem.importMutationEventRaised {
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: Use `getMobileCompatible` since it is an atomic

File: parts/xmem_nozzle.go [hunk 2]
```diff
@@ -755,6 +755,9 @@ type XmemNozzle struct {
 	memcachedErrMap map[string]interface{}
 
 	eventsProducer common.PipelineEventsProducer
+
+	importMutationEventRaised bool
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: Make this an atomic

File: parts/xmem_nozzle.go [hunk 3]
```diff
@@ -1340,6 +1343,14 @@ func (xmem *XmemNozzle) batchSetMetaWithRetry(batch *dataBatch, numOfRetry int)
 					ImportMutation: item.ImportMutation,
 				}
 				xmem.RaiseEvent(common.NewEvent(common.DataFailedCRSource, nil, xmem, nil, additionalInfo))
+				if item.ImportMutation && xmem.config.mobileCompatible == base.MobileCompatibilityOff && !xmem.importMutationEventRaised {
+					xmem.importMutationEventRaised = true
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: Use atomic

File: parts/xmem_nozzle.go [hunk 4]
```diff
@@ -2615,7 +2626,14 @@ func (xmem *XmemNozzle) receiveResponse(finch chan bool, waitGrp *sync.WaitGroup
 						CloneSyncCh:         wrappedReq.ClonedSyncCh,
 					}
 					xmem.RaiseEvent(common.NewEvent(common.DataSent, nil, xmem, nil, additionalInfo))
-
+					if wrappedReq.ImportMutation && xmem.config.mobileCompatible == base.MobileCompatibilityOff && !xmem.importMutationEventRaised {
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: Same as prev comments

File: parts/xmem_nozzle.go [hunk 5]
```diff
@@ -2615,7 +2626,14 @@ func (xmem *XmemNozzle) receiveResponse(finch chan bool, waitGrp *sync.WaitGroup
...
+						xmem.importMutationEventRaised = true
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: Only raise event if compareAndSwap works on moving it from 0 to 1

File: parts/xmem_nozzle.go [hunk 6]
```diff
@@ -3530,6 +3548,10 @@ func (xmem *XmemNozzle) UpdateSettings(settings metadata.ReplicationSettingsMap)
 		if oldMobileCompatible != mobileCompatible {
 			xmem.Logger().Infof("%v updated %v to %v\n", xmem.Id(), MOBILE_COMPATBILE, mobileCompatible)
 		}
+		if mobileCompatible != base.MobileCompatibilityOff && xmem.importMutationEventRaised {
+			xmem.importMutationEventRaised = false
```
Comment (nelio2k) [inline parts/xmem_nozzle.go]: use atomic

File: parts/xmem_nozzle.go [hunk 7]
```diff
@@ -3530,6 +3548,10 @@ func (xmem *XmemNozzle) UpdateSettings(settings metadata.ReplicationSettingsMap)
...
+			xmem.eventsProducer.DismissEvent(int(xmem.importMutationEventId))
```
Comment (nelio2k) [inline parts/xmem_nozzle.go:3550]: Use atomic

File: parts/xmem_nozzle.go [hunk 8]
```diff
@@ -755,6 +755,9 @@ type XmemNozzle struct {
 	memcachedErrMap map[string]interface{}
 
 	eventsProducer common.PipelineEventsProducer
+
+	importMutationEventRaised bool
+	importMutationEventId     int64
```
Comment (nelio2k) [inline parts/xmem_nozzle.go:760]: Use atomic

File: tools/testScripts/testLibrary.shlib [hunk 9]
```diff
@@ -1687,5 +1687,6 @@ function setBucket {
 
 	local port=${CLUSTER_NAME_PORT_MAP[$cluster]:-}
 	echo "$CURL -X POST -u $DEFAULT_ADMIN:$DEFAULT_PW http://127.0.0.1:$port/pools/default/buckets/$bucketName -d $setting=$value"
-	$CURL -X POST -u $DEFAULT_ADMIN:$DEFAULT_PW http://127.0.0.1:$port/pools/default/buckets/$bucketName -d $setting=$value
+	output=$($CURL -X POST -u $DEFAULT_ADMIN:$DEFAULT_PW http://127.0.0.1:$port/pools/default/buckets/$bucketName -d $setting=$value)
```
Comment (nelio2k) [inline tools/testScripts/testLibrary.shlib:1690]: Why do you want to change the output here?

File: parts/xmem_nozzle.go (hunk 1 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

File: parts/xmem_nozzle.go (hunk 4 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

File: tools/testScripts/testLibrary.shlib (hunk 9 above)
Comment (lileichen) [reply tools/testScripts/testLibrary.shlib:1690]: No change to output. Without my change here, any error will be displayed without a return. The next output in the test will continue on the same line and it is confusing.

**Comment by nelio2k** [comment]
While going over old code reviews, can you also change:

`useCasLocking()`
to use the atomic

File: parts/xmem_nozzle.go (hunk 5 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

File: parts/xmem_nozzle.go (hunk 7 above)
Comment (lileichen) [reply parts/xmem_nozzle.go:3550]: Done

File: parts/xmem_nozzle.go (hunk 8 above)
Comment (lileichen) [reply parts/xmem_nozzle.go:760]: Done

File: parts/xmem_nozzle.go (hunk 6 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

File: parts/xmem_nozzle.go (hunk 3 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

File: parts/xmem_nozzle.go (hunk 2 above)
Comment (lileichen) [reply parts/xmem_nozzle.go]: Done

**Comment by nelio2k** [comment]
I don't see the `useCasLocking()` fix here. Did you forget to commit the change?
If you didn't plan on fixing it, let me know and I will file a MB to fix it later.

**Review by nelio2k — APPROVED** [review]

**Pull request merged**